src/app/editor/contextualListener.js

```diff
@@ -0,0 +1,93 @@
+'use strict'
+var SourceMappingDecoder = require('ethereum-remix').util.SourceMappingDecoder
+var AstWalker = require('ethereum-remix').util.AstWalker
+
+class ContextualListener {
+  constructor (api, events) {
+    this._api = api
+    this._index = {
+      ReferencedDeclarations: {},
+      Expressions: []
+    }
+    this._events = []
+
+    events.compiler.register('compilationFinished', (success, data, source) => {
+      this._stopWarning()
+      this._index = {
+        ReferencedDeclarations: {},
+        Expressions: []
+      }
+      if (success) {
+        this._buildIndex(data, source)
+      }
+    })
+    this.sourceMappingDecoder = new SourceMappingDecoder()
+    this.astWalker = new AstWalker()
+    setInterval(() => {
+      this._warnExpressions(api.getCursorPosition(), api.getCompilationResult())
+    }, 1000)
+  }
+
+  _warnExpressions (cursorPosition, compilationResult) {
+    if (this.currentPosition === cursorPosition) return
+    this._stopWarning()
+    this.currentPosition = cursorPosition
+    if (compilationResult && compilationResult.data && compilationResult.source) {
+      var nodes = this.sourceMappingDecoder.nodesAtPosition(null, cursorPosition, compilationResult.data.sources[compilationResult.source.target])
+      if (nodes && nodes.length && nodes[nodes.length - 1]) {
+        this._warnExpression(nodes[nodes.length - 1], compilationResult)
+      }
+    }
+  }
+
+  _buildIndex (compilationResult, source) {
+    if (compilationResult && compilationResult.sources) {
+      var self = this
+      var callback = {}
+      callback['*'] = function (node) {
+        if (node && node.attributes && node.attributes.referencedDeclaration) {
+          if (!self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration]) {
+            self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration] = []
+          }
+          self._index['ReferencedDeclarations'][node.attributes.referencedDeclaration].push(node)
+          self._index['Expressions'].push(node)
```

Comment:
Why is it called "expressions"?

Reply:
in this context, `Expressions` would be a general term to describe use of a declared item.

Reply:
Why not `declaration` then?

Reply:
The two sides of the "link" should be called `declaration` and `reference`. Neither of them has to be an expression.